realm/realm-library/src/main/cpp/java_object_accessor.hpp

```diff
@@ -0,0 +1,320 @@
+    util::Any box(bool v) const { return _impl::JavaClassGlobalDef::new_boolean(m_env, v); }
+    util::Any box(double v) const { return _impl::JavaClassGlobalDef::new_double(m_env, v); }
+    util::Any box(float v) const { return _impl::JavaClassGlobalDef::new_float(m_env, v); }
+    util::Any box(int64_t v) const { return _impl::JavaClassGlobalDef::new_long(m_env, v); }
+    util::Any box(util::Optional<bool> v) const { return v ? _impl::JavaClassGlobalDef::new_boolean(m_env, v.value()) : nullptr; }
+    util::Any box(util::Optional<double> v) const { return v ? _impl::JavaClassGlobalDef::new_double(m_env, v.value()) : nullptr; }
+    util::Any box(util::Optional<float> v) const { return v ? _impl::JavaClassGlobalDef::new_float(m_env, v.value()) : nullptr; }
+    util::Any box(util::Optional<int64_t> v) const { return v ? _impl::JavaClassGlobalDef::new_long(m_env, v.value()) : nullptr; }
+    util::Any box(RowExpr) const { REALM_TERMINATE("'RowExpr' not implemented"); }
+
+    // Mixed type is only supported by the Cocoa binding to enable reading
+    // old Realm files that may have used them. All other bindings can ignore it.
+    util::Any box(Mixed) const { REALM_TERMINATE("'Mixed' not supported"); }
+
+    // Convert from the boxed type to core types. This needs to be implemented
+    // for all of the types which `box()` can take, plus `RowExpr` and optional
+    // versions of the numeric types, minus `List` and `Results`.
+    //
+    // `create` and `update` are only applicable to `unbox<RowExpr>`. If
+    // `create` is false then when given something which is not a managed Realm
+    // object `unbox()` should simply return a detached row expr, while if it's
+    // true then `unbox()` should create a new object in the context's Realm
+    // using the provided value. If `update` is true then upsert semantics
+    // should be used for this.
+    template<typename T>
+    T unbox(util::Any& /*v*/, bool /*create*/= false, bool /*update*/= false) const {
+        throw std::logic_error("Missing template specialization"); // All types should have specialized templates
+    }
+
+    bool is_null(util::Any const& v) const noexcept { return !v.has_value(); }
+    util::Any null_value() const noexcept { return {}; }
+    util::Optional<util::Any> no_value() const noexcept { return {}; }
+
+    // KVO hooks which will be called before and after modying a property from
+    // within Object::create().
+    void will_change(Object const&, Property const&) {}
+    void did_change() {}
+
+    // Get a string representation of the given value for use in error messages.
+    // This method should only be used when printing warnings about primary keys
+    // which means the input should only be valid types for primary keys:
+    // StringData, int64_t and Optional<int64_t>
+    std::string print(util::Any const& val) const {
+        if (!val.has_value()) {
+            return "null";
+        }
+
+        if (val.type() == typeid(StringData)) {
```

Comment:
Isn't there a better way to just print the value than trying to anticipate what kind of types are being sent?
Also, `typeid` doesn't seem to work the same way on different architectures. I'm seeing deterministic crashes don CI that I cannot reproduce locally (which are about these checks all failing so the final logic error is thrown).

Reply:
maybe you need to use `hash_code()` to get a guarantee?

Reply:
How would that work?

Reply:
From https://en.cppreference.com/w/cpp/language/typeid:

"There is no guarantee that the same std::type_info instance will be referred to by all evaluations of the typeid expression on the same type, although std::type_info::hash_code of those type_info objects would be identical"

Reply:
I'm not 100% sure, but https://stackoverflow.com/a/40357969/1389357 seems to indicate that it might not always be the case?